### PR TITLE
Redesign pixelpipe cache importance

### PR DESCRIPTION
**Pull request by jenshannoschwalm**
In the current desing we tried to give modules output buffers some higher cache priority via fixed rules. This works fine in some cases, in other cases this fails miserably. Also Pascal hinted concerns about safety & stability.

So let's rethink the "importance strategy".

In fact this is all about the user interface and the only thing we can take for sure is:

IF the user changed a parameter in a module there is a very good chance that there we be another change of parameters in the same module pretty soon (just think about sliders)

So we want the **input** cacheline of that module to be available afterwards.

We had the strategy of "module in focus" to do so, in this pr we extend that rule by adding a hint in `_dev_add_history_item_ext()`

While using opencl we make sure that these input cachelines are written.

Overall
- a much better cache hit rate
- more respnsiveness in UI
- less code complexity

### CODE REVIEW

**Review by TurboGit — APPROVED** [review]
Looks good to me and found no issue. This implementation is indeed cleaner. Thanks.

**Comment by dterrahe** [comment]
Wow. @ralfbrown and I have been looking at the framing module recently and that seems to be an instance where this makes a big difference.

Is it just me (and my aging 1060) or does 4.4 really feel like a huge improvement in responsiveness over, say, 4.0.

@TurboGit Maybe we should have a first entry in the release notes under The Big Ones stating that performance/responsiveness has been a major focus in this release and that _really_ substantial gains have been made, with a lot of hard work put in and contributions by many people.

**Comment by jenshannoschwalm** [comment]
Yes it does for me too. And - for us midi device users - this IS a big step.

**Comment by todd-prior** [comment]
> Wow. @ralfbrown and I have been looking at the framing module recently and that seems to be an instance where this makes a big difference.
> 
> Is it just me (and my aging 1060) or does 4.4 really feel like a huge improvement in responsiveness over, say, 4.0.
> 
> @TurboGit Maybe we should have a first entry in the release notes under The Big Ones stating that performance/responsiveness has been a major focus in this release and that _really_ substantial gains have been made, with a lot of hard work put in and contributions by many people.

For sure we should be certain to make some noise and give credit. I believe not long ago I made a similar remark in a post on pixls about all this hard work going on in the background to refine and optimize things... no new flashy modules but alot of code refinement.... Well done all of you slaving away on this .....

**Comment by dterrahe** [comment]
> * less code complexity

Maybe so, but still pretty daunting for the uninitiated. So I have no idea how to even find the answer to my next question, so will ask _you_, in the hope you'll not give it more importance than it deserves:

This PR made me wonder if masks and processing outputs are cached separately (not talking about raster masks which are "saved", not "cached" for later use). For example if I'm adjusting the size of a simple circle mask in a heavy diffuse or sharpen module, judging by recalc time it triggers a full recalc each time. Or if I made a very complicated drawn mask in an exposure module, changing just the exposure seems to take slightly longer, presumably because the mask gets redrawn each time? So specifically for the case mentioned here, where subsequent adjustments are made to the same module, when the module output _or_ the mask could be reused as is (this wouldn't work for parametric masks linked to output!) would caching them individually be possible at all? (without upending the whole caching framework; this could in theory just be 1 output and 1 mask for the "current" module).

And separately, but related again, it seems we could quite easily add another _combine_mask mode "additive" (besides union, intersect etc) where the resulting mask = opacity_a * mask_a + opacity_b * mask_b (capped at 1 obviously), rather than MAX of a,b. If we did that, then the frequent dodge and burn requests we get, where people want to be able to "deepen" a mask by brushing on the same spot multiple time, could be quite easily achieved by just drawing with a lower opacity. With the caveat that a brush only gets applied after it is finished, so you would have to lift the pen after each stroke. And you could end up with lots of small (inefficient/expensive) brush shapes, so eventually a good algorithm to merge them (and brushes with variable opacity) would be nice. But if the current mask is cached (as per my first question) then any additional newly created shapes could just be drawn on top of the existing mask and the cached current processing output could be reused as well, so responsiveness could be good (while drawing! frequent undo could ruin it).

This all in theory; in practice this stuff always gets messy/might not be worth it.

**Comment by jenshannoschwalm** [comment]
> This PR made me wonder if masks and processing outputs are cached separately

No.


> not talking about raster masks which are "saved", not "cached" for later use

I wouldn't say they are saved. Yes - internally - but the status (is it written at all and who takes it - the targets) of a raster mask is highly dynamic and only valid after processing the source module.

 

> For example if I'm adjusting the size of a simple circle mask in a heavy diffuse or sharpen module, judging by recalc time it triggers a full recalc each time

This has been the reason i tried the "important hint" idea for a module's output or even pass that hint of importance to the next module. Works nicely in some cases but the bad thing is, it "pollutes" cachelines not available any more in the next pixelpipe run. One of the concepts of the pixelpipe cache is; we may take many cachelines (up to ~60 for full pipe) but have to do some housekeeping to keep memory usage under control. (We could take more mem for the cache but then would have to restrict usage for module processing leading to more tiling.)

> So specifically for the case mentioned here, where subsequent adjustments are made to the same module, when the module output _or_ the mask could be reused as is (this wouldn't work for parametric masks linked to output!) would caching them individually be possible at all?

Good point. I think so but THAT might get pretty daunting :-)

The biggest problems i had with the whole caching stuff were some issues very difficult to get down to because they happen only under certain conditions. (Both raster masks and details masks were affected)
At least we now have much better changes for debugging & testing ...


> And separately, but related again, it seems we could quite easily add another _combine_mask mode "additive" (besides union, intersect etc) where the resulting mask = opacity_a * mask_a + opacity_b * mask_b (capped at 1 obviously), rather than MAX of a,b. If we did that, then the frequent dodge and burn requests we get, where people want to be able to "deepen" a mask by brushing on the same spot multiple time, could be quite easily achieved by just drawing with a lower opacity in "multiple brush strokes" mode (ctrl+).

That really sound easier to be implemented. YET - i have never been using masks as extensively on my photos as while i was working on all this stuff.

  Just one thing completely off-topic. This pixelpipe related and caching code is something i feel my feet are on stable ground. If i see your code - i understand "nothing". Not ranting at all :-) I might be really good to have some sort of meeting where we could introduce & discuss things we learned in dt code. Other examples are ralfs compiler tunings or what pascal did on presets stuff.

**Comment by ralfbrown** [comment]
In my case, I've now spent enough hours looking at disassemblies (objdump -S) to have a fairly intuitive feel for when GCC will produce good vectorized code and when not.  But I'm still wrong an awful lot.  In performance tuning, the only way to be sure a change is actually an improvement is to benchmark it.  As a side effect :-) I get all those numbers I include in my PRs.

**Comment by dterrahe** [comment]
> That really sound easier to be implemented.

Started here #14475
Needs some fixes so that drawing doesn't need to be interrupted to change the mode to sum for each individual brush stroke. On a fast machine with a simple module this might be useful even without the separation of processing and masks caching.

**Comment by dterrahe** [comment]
> It might be really good to have some sort of meeting where we could introduce & discuss things we learned in dt code.

I almost see dt as several separate applications that happen to have crashed into each other and fused. A lot of the code seems to historically have gravitated to smallest common factor (yeah, I'm not saying "greatest" on purpose). When trying to improve anything, automatically the code starts to become less "common". But also, a computational or caching infrastructure is much more under our control, with limitations dictated by backward compatibility, whereas the UI stuff is still closely tied to gtk (because it sets the boundaries of what's possible, if anything) and historical choices until somebody is crazy enough to rip out stuff and attempt a whole intermediate layer (like I did with inputng, ahem). But since we don't have serious enough testing, you can only replace so much at the same time, so you'll still frankenstein your way along until you get somewhere decently "clean" again (but still stuck with choices that you made because you needed a gradual transition and _some_ remaining users).

On the one hand, having more documentation or design specifications would have really helped getting into any subsystem faster. On the other hand I'm convinced that many parts would have changed without documentation or commentary keeping up, so the confusion might even have been higher. So what I think might be most efficient, because I'll never put in the time to learn as much about the stuff you have worked on as you do, is to have early exchanging of ideas (we call them stupid questions) whenever we'd want to improve something in the areas we know nothing about. And to ask for reviews (real ones, with suggestions for even somewhat larger adjustments/cooperation) whenever a PR straddles a "boundary".

In short; don't expect me to sit down and give a presentation on anything (or write documentation to never receive any feedback on), but I'd be very happy to answer (wide ranging) questions.

**Comment by todd-prior** [comment]
As a curioius idiot I follow as best I can your work and that of the others working on issues and the code base. I take from it what I can. I appreciate all of your collective efforts both from the perspective of results and from your comments/commentary  on the challenges and the process as it evolves…..

Your willingness and that of the others to respond and explain things on occasion to said “curious idiot” is appreciated.

Thank you…..
